dartagnan/src/main/java/com/dat3m/dartagnan/program/memory/Memory.java

```diff
@@ -26,36 +35,68 @@ public Location getLocationForAddress(Address address){
         return map.inverse().get(address);
     }
 
-    public BoolExpr encode(Context ctx){
-        BoolExpr enc = ctx.mkTrue();
+    public BooleanFormula encode(SolverContext ctx){
+    	FormulaManager fmgr = ctx.getFormulaManager();
+		BooleanFormulaManager bmgr = fmgr.getBooleanFormulaManager();
+
+		BooleanFormula enc = bmgr.makeTrue();
         for(List<Address> array : arrays.values()){
-            Expr e1 = array.get(0).toZ3Int(ctx);
+        	Formula e1 = array.get(0).toIntFormula(ctx);
+        	boolean bv = e1 instanceof BitvectorFormula;
             for(int i = 1; i < array.size(); i++){
-                Expr e2 = array.get(i).toZ3Int(ctx);
-                Expr newAddress = e1.isBV() ? ctx.mkBVAdd((BitVecExpr)e1, ctx.mkBV(1, array.get(0).getPrecision())) : ctx.mkAdd((IntExpr)e1, ctx.mkInt(1));
-				enc = ctx.mkAnd(enc, ctx.mkEq(e2, newAddress));
+            	Formula e2 = array.get(i).toIntFormula(ctx);
+				Formula newAddress = bv ?
+						fmgr.getBitvectorFormulaManager().add((BitvectorFormula)e1, fmgr.getBitvectorFormulaManager().makeBitvector(array.get(0).getPrecision(), BigInteger.ONE)) :
+							fmgr.getIntegerFormulaManager().add((IntegerFormula)e1, fmgr.getIntegerFormulaManager().makeNumber(BigInteger.ONE));
+				enc = bmgr.and(enc, bv ? 
+						fmgr.getBitvectorFormulaManager().equal((BitvectorFormula)e2, (BitvectorFormula)newAddress) : 
+							fmgr.getIntegerFormulaManager().equal((IntegerFormula)e2, (IntegerFormula)newAddress));
                 e1 = e2;
             }
         }
         // Following SMACK, only address with constant values can have negative values.
         for(Address add : getAllAddresses()) {
         	if(!add.hasConstantValue()) {
-        		enc = ctx.mkAnd(enc, add.toZ3Int(ctx).isBV() ? ctx.mkGt(ctx.mkBV2Int((BitVecExpr) add.toZ3Int(ctx), false), ctx.mkInt(0)) : ctx.mkGt((IntExpr)add.toZ3Int(ctx), ctx.mkInt(0)));
+        		enc = bmgr.and(enc, add.toIntFormula(ctx) instanceof BitvectorFormula ?
+        				fmgr.getIntegerFormulaManager().greaterThan(fmgr.getBitvectorFormulaManager().toIntegerFormula((BitvectorFormula)add.toIntFormula(ctx), false), fmgr.getIntegerFormulaManager().makeNumber(BigInteger.ZERO)) :
+        				fmgr.getIntegerFormulaManager().greaterThan((IntegerFormula)add.toIntFormula(ctx), fmgr.getIntegerFormulaManager().makeNumber(BigInteger.ZERO)));
         	}
         }
         
-        return ctx.mkAnd(enc, ctx.mkDistinct(getAllAddresses().stream().map(a -> a.toZ3Int(ctx).isBV() ? ctx.mkBV2Int((BitVecExpr) a.toZ3Int(ctx), false) : a.toZ3Int(ctx)).toArray(Expr[]::new)));
+        BooleanFormula distinct = getAllAddresses().size() > 1 ?
+        		fmgr.getIntegerFormulaManager().distinct(getAllAddresses().stream()
+                		.map(a -> convertToIntegerFormula(a.toIntFormula(ctx), ctx))
+                		.collect(Collectors.toList())) : 
+                bmgr.makeTrue();
+        
+        return bmgr.and(enc, distinct);        	
```

Comment:
Refactor fmgr.getIntegerFormulaManager into local variable (it will always be used anyway)